zk/src/org/zkoss/zk/ui/select/impl/ComponentIterator.java

```diff
@@ -291,51 +305,118 @@ private ComponentMatchCtx buildNextCtx() {
 			return null;
 
 		// TODO: how to skip tree branches
-
-		// traverse shadow element tree only when selecting shadow elements
-		if (_lookingForShadow && _currCtx.isShadowHost()) {
-			return buildFirstShadowChildCtx(_currCtx);
-		}
-
-		// traverse non shadow component tree
 		if (_currCtx.getComponent().getFirstChild() != null) {
 			return buildFirstChildCtx(_currCtx);
 		}
 
 		while (_currCtx.getComponent().getNextSibling() == null) {
-			if (_lookingForShadow) {
-				ShadowElement se = getNextShadowRootSibling();
-				if (se != null) {
-					return buildNextShadowSiblingCtx(_currCtx, se);
-				}
-			}
 			_currCtx = _currCtx.getParent();
 			if (_currCtx == null || _currCtx.getComponent() == (_posOffset > 0 ? _offsetRoot : _root))
 				return null; // reached root
 		}
 		return buildNextSiblingCtx(_currCtx);
 	}
 
-	// shadow root have no parent, only host. Retrieve sibling from host's seRoots, if there is any.
-	private ShadowElement getNextShadowRootSibling() {
+	private Component getNextUntrackedChild(Component comp) {
+		Component child = comp.getFirstChild();
+		while (child != null) {
+			if (_trackedUuid.contains(child.getUuid())) {
+				child = child.getNextSibling();
+			} else {
+				break; //found
+			}
+		}
+		return child;
+	}
+
+	private Component getNextUntrackedDistributedChild(List<Component> distributedChildren) {
+		for (int i = 0, size = distributedChildren.size(); i < size; i++) {
+			Component comp = distributedChildren.get(i);

```

Comment:
should not use get(i) here, please check how the implementation of getDistributedChildren() is.